### PR TITLE
subdomain matching in policy.json

### PR DESCRIPTION
**Pull request by lsm5**
Signed-off-by: Lokesh Mandvekar <lsm5@fedoraproject.org>

@mtrmac @vrothberg @ashcrow PTAL

### CODE REVIEW

File: docker/policyconfiguration/naming.go
Comment (mtrmac) [inline docker/policyconfiguration/naming.go:65]: Given a `foo.bar`, this turns `name` into `*.bar`, i.e. the same number of dots, so the loop never terminates AFAICS. The tests running out of memory are consistent with that.

**Review by mtrmac — COMMENTED** [review]
Tests are failing (see the Travis output, or during development, run `make` in the checkout).

I’d expect `naming_test.go` to need updating, at the very least.

Given how security-critical this is, tests should aim at 100% code coverage (apart from clearly unreachable code) for the `naming.go` file and for `docker_transport.go`, and at least _some_ “integration” checks in `c/image/signature` that this has the expected effect.

**Comment by lsm5** [comment]
some comments addressed, some yet to be addressed..

File: docker/policyconfiguration/naming.go
Comment (ashcrow) [reply docker/policyconfiguration/naming.go:68]: Could https://golang.org/pkg/net/url/#Parse be helpful here?

File: docker/policyconfiguration/naming.go
Comment (ashcrow) [inline docker/policyconfiguration/naming.go:60]:
```suggestion
        // Iterate over the URI splitting the host by dots into an array
	for {
```

File: docker/policyconfiguration/naming.go
Comment (lsm5) [reply docker/policyconfiguration/naming.go:69]: @ashcrow do you mean rewrite the entire splitting including namespaces/subdomains using url.Parse? Or just for stripping the port number? I used portNumColon for detecting port, but I guess I could use SplitHostPort.

Comment (lsm5) [reply docker/policyconfiguration/naming.go:69]: @ashcrow is this what you meant?

```
-       portNumColon := strings.Index(name, ":")
-       name = name[:portNumColon]
+       HostPort, err := url.Parse(name)
+       if err != nil {
+               panic(err)
+       }
+       name, _, _ = net.SplitHostPort(HostPort.Host)
```

Comment (mtrmac) [reply docker/policyconfiguration/naming.go:69]: Probably not `url.Parse`, because the inputs are not URLs (e.g. no escaping, and not most of the semantics.)

`SplitHostPort` does a fair bit of extra work, which can be ignored for the input that is actually acceptable to `reference.DomainRegexp`, but it also fails on port-less host names; so it would not simplify much.

File: docker/policyconfiguration/naming.go
Comment (mtrmac) [inline docker/policyconfiguration/naming.go:57]: `[:-1]` (if there is no port) is not valid.

File: docker/policyconfiguration/naming.go
Comment (mtrmac) [inline docker/policyconfiguration/naming.go:58]: This should happen only if `name` was changed, otherwise there would be a duplicate value.

… Actually, I’m not at all sure that we should allow non-wildcard $host matching $host:port , historically that did not match. We can safely introduce new semantics for the wildcard syntax, but this might break existing users. So, I’m leaning towards just removing this.

File: docker/policyconfiguration/naming.go
Comment (mtrmac) [inline docker/policyconfiguration/naming.go:60]: Well, this is not splitting “the host into an array”.

File: docker/docker_transport_test.go
Comment (mtrmac) [inline docker/docker_transport_test.go:151]: docs/containers-policy.json.5.md should eventually be updated.

**Comment by lsm5** [comment]
> From the tests:
> 
> > [f9f19d7](https://github.com/containers/image/commit/f9f19d75cb5cb2358e4fdc9f96f71becf6bcb1ea) - FAIL - does not have a valid DCO
> 
> How did accepting a GitHub suggestion end up with “Co-authored-by: Stephen Milner [stevem@gnulinux.net](mailto:stevem@gnulinux.net)”?

I hit `commit suggestion` and looks like that's what it ended up doing. Anyway, I'll be squashing commits, so the DCO issue will go away.

Still working on other review comments, specially the tests, so I'll update this PR soon. I'll ping you once it's ready for further review and comments.

Thanks

File: docker/policyconfiguration/naming.go
Comment (ashcrow) [reply docker/policyconfiguration/naming.go:69]: @lsm5 yes, but it was more specifically a question as if it would help in this situation. I defer to @mtrmac's note.

**Comment by mtrmac** [comment]
> How did accepting a GitHub suggestion end up with “Co-authored-by: Stephen Milner [stevem@gnulinux.net](mailto:stevem@gnulinux.net)”?

Oops, this actually does match the original author; I have confused the GitHub handle with someone else. I apologize.

File: docker/policyconfiguration/naming.go
Comment (lsm5) [reply docker/policyconfiguration/naming.go:58]: I haven't pushed the change yet, but locally I have:

```
        // Needed for TestReferencePolicyConfigurationNamepsaces
        portNumColon := strings.Index(name, ":")
        if portNumColon != -1 {
                name = name[:portNumColon]
                res = append(res, name)
        }
```

`openshift.TestReferencePolicyConfigurationNamepsaces` uses a port number, and it will fail if I include `*.example.com` and `*.com` but will pass on `*.example.com:8443` and `*.com:8443` which isn't something we want IIUC. So, do we keep this, or do we get rid of port number from the test?

Comment (mtrmac) [reply docker/policyconfiguration/naming.go:58]: I think the most compatible behavior is to return `registry.example.com:8443/ns/stream`, `registry.example.com:8443/ns`, `registry.example.com:8443,`*.example.com`, `*.com` — and not `registry.example.com`.

I can see the argument for including `registry.example.com` anyway, even if it changes behavior of existing configurations, for more consistency. It’s a tough call.

Comment (lsm5) [reply docker/policyconfiguration/naming.go:58]: > I think the most compatible behavior is to return `registry.example.com:8443/ns/stream`, `registry.example.com:8443/ns`, `registry.example.com:8443`,`*.example.com`, `*.com` — and not `registry.example.com`.

Hmm, I guess I will go with this for now. Let me know if you later prefer the one below.

> 
> I can see the argument for including `registry.example.com` anyway, even if it changes behavior of existing configurations, for more consistency. It’s a tough call.

Comment (lsm5) [reply docker/policyconfiguration/naming.go:58]: so if there is no port number mentioned, for example: `registry.example.com/ns/stream`, should the returned array include `registry.example.com`  in that case? or just `[registry.example.com/ns/stream registry.example.com/ns *.example.com *.com]`

Comment (mtrmac) [reply docker/policyconfiguration/naming.go:58]: Yes; it is included now, so it must be included in the updated version as well.

**Comment by lsm5** [comment]
@mtrmac ping, looks like a few tests are failing on current origin/master with `key "1D8230F6CDB6A06716E414C1DB72F2188BB46CC8" not found`. Or is it just me?

**Comment by mtrmac** [comment]
> @mtrmac ping, looks like a few tests are failing on current origin/master with `key "1D8230F6CDB6A06716E414C1DB72F2188BB46CC8" not found`. Or is it just me?

@lsm5 works for me, I think (At least `go test ./copy ./signature ` works for me, and the Travis tests fail differently ; you didn’t say what is failing).

Maybe try killing gpg-agent and/or logging in again.

**Comment by lsm5** [comment]
> > @mtrmac ping, looks like a few tests are failing on current origin/master with `key "1D8230F6CDB6A06716E414C1DB72F2188BB46CC8" not found`. Or is it just me?
> 
> @lsm5 works for me, I think (At least `go test ./copy ./signature ` works for me, and the Travis tests fail differently ; you didn’t say what is failing).
> 
> Maybe try killing gpg-agent and/or logging in again.

ack, I rebooted and I don't see it anymore. Thanks!

**Comment by lsm5** [comment]
@mtrmac PTAL when you get the chance, tests pass now (but rerunning because of rebase)

File: docs/containers-policy.json.5.md
Comment (mtrmac) [inline docs/containers-policy.json.5.md:35]: This should go into the “## Supported transports and their scopes” section, it’s not an universal mechanism (the list above basically matches `requirementsForImageRef`)

File: docs/containers-policy.json.5.md
Comment (mtrmac) [inline docs/containers-policy.json.5.md:234]: I don’t think this scope would be matched

File: signature/docker_test.go
Comment (mtrmac) [inline signature/docker_test.go:27]: Probably drop this

File: signature/policy_eval.go
Comment (mtrmac) [inline signature/policy_eval.go:145]: As far as the generic code cares, wildcards are just parent PolicyConfigurationNamespaces.

**Review by mtrmac — COMMENTED** [review]
A quick pass, I’ll need to think a bit more about the added tests vs. what is needed.

**Comment by lsm5** [comment]
@mtrmac PTAL at the changes. Thanks.

**Comment by lsm5** [comment]
@mtrmac ping, PTAL

**Comment by ashcrow** [comment]
The updates look good to me! @mtrmac thoughts?

**Comment by rhatdan** [comment]
Needs a rebase.

File: docker/daemon/daemon_transport_test.go
Comment (mtrmac) [inline docker/daemon/daemon_transport_test.go:221]: Add this to `TestTransportValidatePolicyConfigurationScope` as well, please.

File: docker/policyconfiguration/naming_test.go
Comment (mtrmac) [inline docker/policyconfiguration/naming_test.go:53]: (Nit: This comment really belongs _inside_ the `if` branch. I’m not sure that it is very necessary either way.

Here, a comment to the effect of “verify that `expectedNS` is ordered from most specific to least specific” might be useful to explain the _purpose_ of the code, and that might sufficiently explain thy the subdomain does what it does.)

File: docker/policyconfiguration/naming_test.go
Comment (mtrmac) [inline docker/policyconfiguration/naming_test.go:55]: Non-blocking: Wouldn’t `[1:]` be a more precise match here?

File: docs/containers-policy.json.5.md
Comment (mtrmac) [inline docs/containers-policy.json.5.md:94]: This needs to be precise enough to allow users to construct that expression (e.g. what the wildcard is), and not suggest that things like `foo*.*.bar` are valid.

Is the extra indentation intentional? It usually has semantic meaning in Markdown.

File: openshift/openshift_transport_test.go
Comment (mtrmac) [inline openshift/openshift_transport_test.go:116]: Add this to `TestTransportValidatePolicyConfigurationScope` as well, please.

File: docker/policyconfiguration/naming.go
Comment (mtrmac) [reply docker/policyconfiguration/naming.go:65]: Now that there is a good comment above, this can just be dropped.

**Review by mtrmac — COMMENTED** [review]
I’m sorry about taking so long.

Code looks good, clear documentation is the only pressing concern, other than that just a few implementation-private cleanups.

The current set of tests is as far as we can reasonably get with c/image. (The Skopeo integration tests could be extended with a minimal smoke test, but without a DNS infrastructure to manipulate it would be cumbersome to test much more, and anyway the existing set of tests should give us enough confidence that I don’t think testing this in Skopeo is necessary at all.)

**Comment by lsm5** [comment]
@mtrmac thanks a lot for the review and comments. I'll quite likely address these along with fixing test failures (post-rebase) after I'm back from break (or maybe during the break if I'm bored). See you all soon..

**Comment by lsm5** [comment]
@mtrmac PTAL, all comments should be addressed with the latest change. Thanks a lot!

**Comment by lsm5** [comment]
also updated the commit message a bit.

File: docs/containers-policy.json.5.md
Comment (mtrmac) [inline docs/containers-policy.json.5.md:95]: `atomic:` above (c/image/openshift) also needs updating. (Possibly also `docker-daemon:`, but that’s currently not documented here at all, so N/A.)

File: docs/containers-policy.json.5.md
Comment (mtrmac) [inline docs/containers-policy.json.5.md:259]: It’s not ideal to recommend insecure use of signed RH images. Something like `*.temporary-project.example.com`? Or maybe show how to require RH signatures for all of `*.redhat.com` (but that could detract from the `*.` point).

**Review by mtrmac — COMMENTED** [review]
Code LGTM, a few more documentation nits. I should have noticed the `atomic:` one before, I’m sorry.

**Comment by TomSweeneyRedHat** [comment]
LGTM once @mtrmac's comments are addressed.

File: docs/containers-policy.json.5.md
Comment (lsm5) [reply docs/containers-policy.json.5.md:95]: Ack, so I could just copy-paste that paragraph on general scopes as-is to the atomic: section, correct? Or, are there any atomic specific changes required there?

Comment (mtrmac) [reply docs/containers-policy.json.5.md:95]: Yes; the two differ slightly in the after-hostname part (and the mandatory `//` prefix) only.

Comment (lsm5) [reply docs/containers-policy.json.5.md:100]: pushed a new change, PTAL. Thanks

Comment (lsm5) [reply docs/containers-policy.json.5.md:100]: ah, noticed your comment only now. I need to change this, sorry i just copy-pasted it earlier.

Comment (lsm5) [reply docs/containers-policy.json.5.md:100]: @mtrmac re-read your comment just now. So, if the difference is only in the prefix and after-hostname, we should be good to go afaict. Let me know, thanks!

File: docs/containers-policy.json.5.md
Comment (mtrmac) [inline docs/containers-policy.json.5.md:263]: Update the comment as well, please.

File: docs/containers-policy.json.5.md
Comment (mtrmac) [inline docs/containers-policy.json.5.md:74]: - The `*note*` paragraph should probably be after, not in the middle of, the scope discussion.
- Hum, the more detailed scopes are “image streams”, not “repositories” — and already described above; so, maybe just continue the “host namespace/image stream” sentence with “, or a wildcarded…”.

File: docs/containers-policy.json.5.md
Comment (lsm5) [reply docs/containers-policy.json.5.md]: whoops, done!

File: docs/containers-policy.json.5.md
Comment (lsm5) [reply docs/containers-policy.json.5.md]: done

**Review by mtrmac — APPROVED** [review]
LGTM. Thanks!

**Comment by rhatdan** [comment]
Is this something we should get into podman 3.0?  Or is it mainly for CRI-O?

**Comment by mtrmac** [comment]
Primary is OpenShift cluster run-time, i.e. CRI-O. Podman is IIRC involved only in parts of the installation process, and I’m not immediately sure whether this is exposed there and will need to be supported.

Either way, I don’t think it is a reason to delay Podman: there will need to be OpenShift updates to accept (and document) the wildcard syntax, it’s not enabled for supported use just by shipping a new CRI-O.

**Pull request merged**
